kafka-connect-upload-source/src/main/java/org/radarbase/connect/upload/converter/xml/XmlProcessor.kt

```diff
@@ -0,0 +1,105 @@
+/*
+ *  Copyright 2019 The Hyve
+ *
+ *  Licensed under the Apache License, Version 2.0 (the "License");
+ *  you may not use this file except in compliance with the License.
+ *  You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS,
+ *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *  See the License for the specific language governing permissions and
+ *  limitations under the License.
+ */
+
+package org.radarbase.connect.upload.converter.xml
+
+import org.radarbase.connect.upload.api.ContentsDTO
+import org.radarbase.connect.upload.api.RecordDTO
+import org.radarbase.connect.upload.converter.ConverterFactory
+import org.radarbase.connect.upload.converter.FileProcessor
+import org.radarbase.connect.upload.converter.TimeFieldParser
+import org.radarbase.connect.upload.converter.TopicData
+import org.w3c.dom.Element
+import java.io.IOException
+import java.io.InputStream
+import javax.xml.parsers.DocumentBuilderFactory
+
+/**
+ * Base class to process XML files with multiple possible per-node/tag processors.
+ */
+open class XmlProcessor(
+        private val record: RecordDTO? = null,
+        private val processorFactories: List<XmlNodeProcessorFactory>? = emptyList(),
+): FileProcessor {
+    val fileNameSuffixes: List<String>
+        get() = listOf(fileNameSuffix)
+
+    open val fileNameSuffix: String = ".xml"
+
+    open val timeFieldParser: TimeFieldParser = TimeFieldParser.EpochMillisParser()
+
+    override fun processData(
+            context: ConverterFactory.ContentsContext,
+            inputStream: InputStream,
+            produce: (TopicData) -> Unit,
+    ) {
+        val factory = DocumentBuilderFactory.newInstance()
+        val builder = factory.newDocumentBuilder()
+
+        if (context.contents.size == 0L) {
+            throw IOException("Cannot read empty XML file ${context.fileName}")
+        }
+
+        val doc: Element = builder.parse(inputStream).documentElement
+        val processors = processorFactories!!.filter { it.matches(context.contents) }.createNodeProcessors(context)
+        val topicDataList = mutableListOf<TopicData>()
+
+        processXml(doc, topicDataList, context, processors, "")
```

Comment:
Why not pass the `produce` function? This way, there’s no need to create additional list and processing is correctly paused when the queue is full.